### PR TITLE
in_exec: allow for long-running commands

### PR DESCRIPTION
**Pull request by j0057**
Add a boolean option (follow) to in_exec plugin to read the spawned
process output using non-blocking I/O, this way Fluent-bit can capture
the output of commands that never exit, for example `vmstat -S M 10`.

Signed-off-by: Joost Molenaar <jjm@j0057.nl>

### CODE REVIEW

**Comment by edsiper** [comment]
cc: @nokute78

File: plugins/in_exec/in_exec.c [hunk 1]
```diff
@@ -119,15 +207,40 @@ static int in_exec_config_read(struct flb_in_exec_config *exec_config,
...
+        exec_config->argv = flb_realloc(exec_config->argv, (argc + 2) * sizeof(char *));
+        exec_config->argv[argc++] = arg;
+    }
+    exec_config->argv[argc] = NULL;
+
+    /* follow setting */
+    follow = flb_input_get_property("follow", in);
+    if (follow != NULL) {
+        if (strcasecmp(follow, "true") == 0) {
+            exec_config->follow = 1;
```
Comment (nokute78) [inline plugins/in_exec/in_exec.c]: We defined FLB_TRUE/FLB_FALSE at flb_macros.h. Would you replace it?

File: plugins/in_exec/in_exec.c [hunk 2]
```diff
@@ -119,15 +207,40 @@ static int in_exec_config_read(struct flb_in_exec_config *exec_config,
...
+    /* split into an argv array */
+    char *arg = NULL;
+    for (arg = strtok_r(cmd, " ", &cmd); arg != NULL; arg = strtok_r(NULL, " ", &cmd)) {
+        exec_config->argv = flb_realloc(exec_config->argv, (argc + 2) * sizeof(char *));
+        exec_config->argv[argc++] = arg;
+    }
+    exec_config->argv[argc] = NULL;
+
+    /* follow setting */
+    follow = flb_input_get_property("follow", in);
```
Comment (nokute78) [inline plugins/in_exec/in_exec.c]: Refer to in_exec of fluentd, `run_interval` option means `follow` mode or not.
https://docs.fluentd.org/v1.0/articles/in_exec#run_interval

I means, 
I think we don't need new option "follow".

If `Interval_Sec` and `Interval_NSec` are not set, `exec_config->follow` is FLB_TRUE. (= run_interval is not set)
if `Interval_Sec` or/and `Interval_NSec` is/are set, `exec_config->follow` is FLB_FALSE. (= run_interval is set)
Default value of `Interval_Sec` and `Interval_NSec` are 0.

Comment (nokute78) [reply plugins/in_exec/in_exec.c]: Note: We support fd event based API `flb_input_set_collector_event`. (e.g. in_kmsg uses this API)

I think the follow mode is not time based mode. How about trying to use it?

```c
flb_input_set_collector_event(in,
                                           in_exec_collect,
                                           fileno(exec_config->cmdp),
                                           config);
```

**Comment by nokute78** [comment]
@j0057 I commented on your commit.

Defunct process will be generated if user set such configuraion.
(Follow mode and command which exits immediately)
Do you have any idea?
```c
[INPUT]
    Name          exec
    Tag           exec_ls
    Command       ls  /var/log
    follow true

[OUTPUT]
    Name   stdout
    Match  *
```

**Comment by edsiper** [comment]
@j0057 ping

File: plugins/in_exec/in_exec.c (hunk 1 above)
Comment (j0057) [reply plugins/in_exec/in_exec.c]: I updated the code to make use of the FLB_TRUE/FLB_FALSE macros.

**Comment by edsiper** [comment]
@nokute78 is this ready to go after the latest changes or it needs some changes ?

**Comment by edsiper** [comment]
@nokute78 ping

**Comment by edsiper** [comment]
thanks @nokute78 

@j0057 please advice if you want to include the improvement with the changes suggested.

**Comment by edsiper** [comment]
Thanks anyways for the effort on this, no worries, it's just code :) no waste of time, have a great 2019!

@nokute78 thank you for reviewing and help on this

**Pull request closed without merging**